### ISSUE TITLE
Concurrent static initialization of ClassName and TypeName can cause a deadlock

### ISSUE THREAD

**Issue by millems** [square/javapoet#799]
`ClassName` and `TypeName`'s static initialization has a mutual dependency, which can cause a deadlock when they are statically initialized concurrently.

The following test *sometimes* fails:

```java
public class Repro {
    public static void main(String... args) throws Exception {
        Thread classNameLoadingThread = new Thread(() -> ClassName.get(Object.class));
        Thread typeNameLoadingThread = new Thread(() -> TypeName.get(Object.class));

        classNameLoadingThread.start();
        typeNameLoadingThread.start();

        classNameLoadingThread.join(5_000);
        typeNameLoadingThread.join(5_000);

        if (classNameLoadingThread.isAlive() || typeNameLoadingThread.isAlive()) {
            throw new AssertionError("Threads deadlocked!");
        }
    }
}
```

We can make the test fail more reliably by repeating the static initalization multiple times:

```java
public class Repro {
    public static void main(String... args) throws Exception {
        URL[] classesToTest = {urlOf(ClassName.class), urlOf(TypeName.class)};
        for (int i = 0; i < 10000; ++i) {
            ClassLoader classLoader = new URLClassLoader(classesToTest);
            Thread classNameLoadingThread = new Thread(() -> loadClass(classLoader, ClassName.class));
            Thread typeNameLoadingThread = new Thread(() -> loadClass(classLoader, TypeName.class));

            classNameLoadingThread.start();
            typeNameLoadingThread.start();

            classNameLoadingThread.join(5_000);
            typeNameLoadingThread.join(5_000);

            if (classNameLoadingThread.isAlive() || typeNameLoadingThread.isAlive()) {
                throw new AssertionError("Threads deadlocked!");
            }
        }
    }

    private static void loadClass(ClassLoader classLoader, Class<?> clazz) {
        try {
            Class<?> aClass = classLoader.loadClass(clazz.getName());
            aClass.getField("OBJECT").get(null);
        } catch (Throwable e) {
            throw new IllegalStateException(e);
        }
    }

    private static URL urlOf(Class<?> clazz) throws Exception {
        return clazz.getResource('/' + clazz.getName().replace('.', '/') + ".class").toURI().toURL();
    }
}
```

An example of the deadlocked threads can be seen here:

```
"Thread-1" #13 prio=5 os_prio=31 cpu=1.81ms elapsed=5.77s tid=0x00007f9e568c7000 nid=0x6603 in Object.wait()  [0x000070000a678000]
   java.lang.Thread.State: RUNNABLE
	at com.squareup.javapoet.TypeName.<clinit>(TypeName.java:78)
	- waiting on the Class initialization monitor for com.squareup.javapoet.ClassName
	at Repro.lambda$main$1(SpeedUpCodegen.java:35)
	at Repro$$Lambda$16/0x0000000800b42040.run(Unknown Source)
	at java.lang.Thread.run(java.base@13.0.1/Thread.java:830)
```

```
"Thread-0" #12 prio=5 os_prio=31 cpu=10.43ms elapsed=5.78s tid=0x00007f9e5612f800 nid=0xa103 in Object.wait()  [0x000070000a575000]
   java.lang.Thread.State: RUNNABLE
	at Repro.lambda$main$0(SpeedUpCodegen.java:34)
	- waiting on the Class initialization monitor for com.squareup.javapoet.TypeName
	at Repro$$Lambda$15/0x0000000800b41440.run(Unknown Source)
	at java.lang.Thread.run(java.base@13.0.1/Thread.java:830)
```

I can workaround this by making sure to statically initialize `ClassName` before branching out to other threads, so it's not urgent to fix this, personally.

**Comment by octylFractal** [comment]
Duplicate of #637.
